### ISSUE TITLE
DL4J: Move word vectors off-heap for WordVectorSerializer.loadStaticModel

### ISSUE THREAD

**Issue by AlexDBlack** [eclipse/deeplearning4j#6721]
I thought there was already an issue for this, but I can't find it.
Here's what the heap looks like with loading google news vectors via loadStaticModel, with nothing else in the JVM:
![image](https://user-images.githubusercontent.com/2360237/48590757-de5f0180-e994-11e8-9bcd-b43a6e71400e.png)
That's a total of 67.4 million objects for a vocab of 3 million words.

One option here: https://github.com/OpenHFT/Chronicle-Map
If we used chronicle map, it would store vectors off-heap.
I was thinking we have a few options here for the internal representation
* ```Map<String,Integer>``` plus ```Map<Integer,INDArray>``` (integers being the index)
* ```Map<String,Integer>``` plus ```Map<String,INDArray>```
* ```Map<String,Pair<Long,Long>>``` plus a float pointer (long values are offsets for each word). Might be faster to load due to less preprocessing.

There's also the idea of using a trie instead of a ```Map<String,X>``` as that would give us reduced memory for storing the strings.
https://en.wikipedia.org/wiki/Trie

Edit: another thought here is that we often want to do 'batch' lookup.
It probably makes sense to do batch lookup and create an INDArray in a single op, rather than do N lookups + combine manually in Java.
So storage.get("word1", "word2", "word3") would return an INDArray with shape [3,300] - rather than 3x get ops ([1,300]) + Nd4j.vstack.

Aha! Link: https://skymindai.aha.io/features/DL4J-48

**Comment by AlexDBlack** [comment]
Memory mapped files might have a role to play for the underlying storage.
The main thing here is to get rid of all the on-heap objects (Strings and INDArrays etc) while still providing for O(1) lookup by String and by index.

**Comment by AlexDBlack** [comment]
Quick experiment with ChroniceMap: I was hoping this could be an easy drop-in replacement for the current map-based storage, looks promising.
https://gist.github.com/AlexDBlack/c64b5b874fe2668563d04d3fc971ddc9
```
Insertion speed: INDArrays/sec
115000 - 10739.633918565558/sec (overall) - 10752.68817204301/sec (last 100)
116000 - 10740.74074074074/sec (overall) - 10869.565217391304/sec (last 100)
...
2991000 - 17413.427726415313/sec (overall) - 19607.843137254902/sec (last 100)
2992000 - 17413.977743632724/sec (overall) - 19230.76923076923/sec (last 100)
```
It took about 172 sec to do the conversion on my system... limiting factor is probably the serialization cost.
Note it's critical to disable GC here, otherwise GC overhead is a major performance issue.

**Comment by dgarcia-ccc** [comment]
@AlexDBlack Did you make any progress on this?. I'm quite interested.

**Comment by raver119** [comment]
Yes, we've implemented all required pre-requisites for this. Next step will be actual vocab re-implementation with native bits within.

**Comment by agibsonccc** [comment]
Closing — references a removed/rewritten subsystem (DataVec, NLP/Word2Vec, Model Zoo, Spark, UI, or Arbiter). These have been fundamentally rewritten or removed in the upcoming release. Please open a new issue if you experience similar problems with the new version.
